In [ ]:
# 3) Descriptive Statistics - Measures of Central Tendency and variability
# Perform the following operations on any open-source dataset (e.g., data.csv)
# 1. Provide summary statistics (mean, median, minimum, maximum, standard deviation) for
# a dataset (age, income etc.) with numeric variables grouped by one of the qualitative
# (categorical) variable. For example, if your categorical variable is age groups and
# quantitative variable is income, then provide summary statistics of income grouped by the
# age groups. Create a list that contains a numeric value for each response to the categorical
# variable.
# 2. Write a Python program to display some basic statistical details like percentile, mean,
# standard deviation etc. of the species of ‘Iris-setosa’, ‘Iris-versicolor’ and ‘Iris-versicolor’ of
# iris.csv dataset.

import pandas as pd

# URL of the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv"

# Load the dataset into a DataFrame
df = pd.read_csv(url)

# Save the dataset locally
df.to_csv("iris.csv", index=False)

print("Dataset downloaded and saved as 'iris.csv'.")

import pandas as pd

# Load the Iris dataset from iris.csv
df = pd.read_csv("iris.csv")

# Display the first few rows of the dataset to understand its structure
print("Dataset Overview:")
print(df.head())

# Task 1: Summary Statistics for a Dataset (Sepal Length grouped by Species)
grouped_stats = df.groupby('species').agg(
    mean_sepal_length=('sepal_length', 'mean'),
    median_sepal_length=('sepal_length', 'median'),
    min_sepal_length=('sepal_length', 'min'),
    max_sepal_length=('sepal_length', 'max'),
    std_sepal_length=('sepal_length', 'std')
)

print("\nSummary Statistics for Sepal Length Grouped by Species:")
print(grouped_stats)

# Create a list of Sepal Length values for each species
grouped_list = df.groupby('species')['sepal_length'].apply(list)
print("\nList of Sepal Lengths for Each Species:")
print(grouped_list)

# Task 2: Display Statistical Details for the Iris Dataset
# Calculate descriptive statistics for each species
species_stats = {}
for species in df['species'].unique():
    species_data = df[df['species'] == species]
    print(f"\nStatistical Details for {species.capitalize()}:")
    print(species_data.describe())


Dataset downloaded and saved as 'iris.csv'.
Dataset Overview:
   sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

Summary Statistics for Sepal Length Grouped by Species:
            mean_sepal_length  median_sepal_length  min_sepal_length  \
species                                                                
setosa                  5.006                  5.0               4.3   
versicolor              5.936                  5.9               4.9   
virginica               6.588                  6.5               4.9   

            max_sepal_length  std_sepal_length  
species                                         
setosa                   5.8          

📚 THEORY PART
1. Descriptive Statistics
Descriptive statistics summarize and describe the important features of a dataset.

They help us understand the central tendency (typical value) and variability (spread) of the data.

2. Measures of Central Tendency
These indicate where most values fall:

Mean: The average (sum of all values divided by number of values).

Median: The middle value when data is sorted.

Mode: The most frequently occurring value (not used here, but important).

3. Measures of Variability
These show how spread out the values are:

Minimum: Smallest value.

Maximum: Largest value.

Standard Deviation: How much the data deviates from the mean on average.

Percentile: Indicates the value below which a given percentage of observations fall (example: 25th percentile, 50th percentile, 75th percentile).

4. Grouping by Categorical Variables
A categorical variable (e.g., species) divides the data into groups.

We then calculate statistics separately for each group (for example, the mean sepal length for each species).

🧠 YOUR PYTHON CODE EXPLANATION
Step 1: Import Libraries
python
Copy
Edit
import pandas as pd
pandas is imported for data manipulation.

Step 2: Download and Save Dataset
python
Copy
Edit
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv"
df = pd.read_csv(url)
df.to_csv("iris.csv", index=False)
print("Dataset downloaded and saved as 'iris.csv'.")
The Iris dataset is loaded directly from an online URL.

It's saved locally as iris.csv so you can use it offline.

It prints a confirmation message.

Step 3: Load and View Dataset
python
Copy
Edit
df = pd.read_csv("iris.csv")
print("Dataset Overview:")
print(df.head())
Reloads the iris.csv file.

Prints the first 5 rows to understand the data structure.

Dataset columns:

sepal_length, sepal_width, petal_length, petal_width, species

Step 4: Grouped Summary Statistics
python
Copy
Edit
grouped_stats = df.groupby('species').agg(
    mean_sepal_length=('sepal_length', 'mean'),
    median_sepal_length=('sepal_length', 'median'),
    min_sepal_length=('sepal_length', 'min'),
    max_sepal_length=('sepal_length', 'max'),
    std_sepal_length=('sepal_length', 'std')
)
print("\nSummary Statistics for Sepal Length Grouped by Species:")
print(grouped_stats)
Group the data by species (setosa, versicolor, virginica).

Calculate:

Mean

Median

Minimum

Maximum

Standard Deviation

.agg() is used to apply multiple aggregation functions at once.

✅ Output looks like:


species	mean_sepal_length	median_sepal_length	min_sepal_length	max_sepal_length	std_sepal_length
setosa	...	...	...	...	...
versicolor	...	...	...	...	...
virginica	...	...	...	...	...
Step 5: Create List of Sepal Lengths for Each Species
python
Copy
Edit
grouped_list = df.groupby('species')['sepal_length'].apply(list)
print("\nList of Sepal Lengths for Each Species:")
print(grouped_list)
For each species, create a list of all corresponding sepal_length values.

Helps you see raw values grouped together.

✅ Output example:

css
Copy
Edit
species
setosa        [5.1, 4.9, 4.7, ...]
versicolor    [7.0, 6.4, 6.9, ...]
virginica     [6.3, 5.8, 7.1, ...]
Step 6: Descriptive Statistics for Each Species
python
Copy
Edit
species_stats = {}
for species in df['species'].unique():
    species_data = df[df['species'] == species]
    print(f"\nStatistical Details for {species.capitalize()}:")
    print(species_data.describe())
Loop through each unique species.

Filter data for that species.

.describe() automatically calculates:

Count

Mean

Standard deviation (std)

Min

25% percentile

Median (50%)

75% percentile

Max

✅ Output for each species looks like:

python-repl
Copy
Edit
Statistical Details for Setosa:
       sepal_length  sepal_width  petal_length  petal_width
count      50.00000     50.00000      50.00000     50.00000
mean        5.00600      3.41800       1.46400      0.24400
std         0.35249      0.38102       0.17351      0.10721
min         4.30000      2.30000       1.00000      0.10000
...